examples/albert/run_first_peer.py

```diff
@@ -1,51 +1,158 @@
+
 
 logger = get_logger(__name__)
 
+
+@dataclass
+class CoordinatorArguments(CollaborationArguments):
+    address: Optional[str] = None  # this machine's network address. Use public IP for global experiments,
+    # local address for private runs
+    refresh_period: float = 30  # coordinator will fetch keys from DHT once in this many seconds
+    wandb_project: Optional[str] = None  # learning curves will be published there
+    save_checkpoint_step_interval: int = 5  # coordinator will load and save state from peers once every that many steps
+    upload_model_as: Optional[str] = None  # coordinator will upload the checkpoint to that HuggingFace repo
+    upload_interval: Optional[float] = None  # coordinator will upload model once in this many seconds
+    # Note: You might want to have several initial peers so that if one dies,
+    # new workers still can join the collaboration via alive initial peers' addresses.
+    # Specify initial_peers argument for that purpose
+
+
+class CheckpointHandler:
+    def __init__(self, coordinator_args: CoordinatorArguments, dataset_args: DatasetArguments, dht: hivemind.DHT):
+        self.save_checkpoint_step_interval = coordinator_args.save_checkpoint_step_interval
+        self.upload_model_as = coordinator_args.upload_model_as
+        self.upload_interval = coordinator_args.upload_interval
+        self.previous_step = -1
+
+        config = AlbertConfig.from_pretrained(dataset_args.config_path)
+        self.model = AlbertForPreTraining(config)
+
+        collaboration_args_dict = asdict(coordinator_args)
+        collaboration_args_dict.pop('address')
+        collaboration_args_dict.pop('refresh_period')
+        collaboration_args_dict.pop('wandb_project')
+        collaboration_args_dict.pop('save_checkpoint_step_interval')
+        collaboration_args_dict.pop('upload_model_as')
+        collaboration_args_dict.pop('upload_interval')
+
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in self.model.named_parameters() if not any(nd in n for nd in no_decay)],
+                "weight_decay": 0.01,
+            },
+            {
+                "params": [p for n, p in self.model.named_parameters() if any(nd in n for nd in no_decay)],
+                "weight_decay": 0.0,
+            },
+        ]
+
+        opt = Lamb(
+            optimizer_grouped_parameters,
+            lr=0.00176, weight_decay=0.01, clamp_value=10000.0, debias=True,
+        )
+
+        adjusted_target_batch_size = collaboration_args_dict.pop('target_batch_size') - \
+            collaboration_args_dict.pop('batch_size_lead')
+
+        collaboration_args_dict.pop('initial_peers')
+        collaboration_args_dict.pop('trainer_uuid')
+        collaboration_args_dict.pop('dht_listen_on')
+        collaboration_args_dict.pop('statistics_expiration')
+        collaboration_args_dict.pop('endpoint')
```

Comment:
Dropping all these parsed arguments leads me to the thought that their hierarchy can be improved. Is it possible to extract all common arguments for peers and collaborators to one argument group and then inherit from it in two separate classes?